### PR TITLE
fix(sea-builder): verify downloaded Node.js archives before caching

### LINKED ISSUE #57: sea-builder downloads Node.js archives unverified and can permanently poison its cache
## Background

`@kurone-kito/sea-builder` downloads official Node.js runtime archives
and hands them to `xsea` to produce distributable single-file
executables. The download path is the supply chain for every binary the
tool emits, and it currently performs no validation at all.

`packages/sea-builder/src/utils/downloadArchive.mts`:

```ts
export const downloadArchive: DownloadFunction = async (url, dest) => {
  const response = await fetch(url);
  if (!response.body) {
    throw new Error('Response body is empty');
  }
  // @ts-expect-error
  await pipeline(response.body, createWriteStream(dest));
};
```

Three defects compound here:

1. **No HTTP status check.** A non-2xx response still has a body, so a
   404 page or a proxy error page is streamed into `dest` and stored
   under the archive's name.
2. **No integrity verification.** `nodejs.org/dist/<version>/` publishes
   `SHASUMS256.txt` alongside every archive precisely so consumers can
   verify what they downloaded. Nothing here reads it.
3. **Non-atomic write into the permanent cache.** The stream writes
   directly to the final path. `src/tasks/createTaskFactory.mts` decides
   whether to download using `existsSync(archivePath)`, so any file left
   behind by an interrupted or failed download — a truncated archive, an
   HTML error page — is treated as a valid cache hit **forever**. The
   documented recovery in `packages/sea-builder/README.md` is for a user
   to notice and delete `node_modules/.cache/xsea` by hand.

The consequence is that a transient network failure or a single bad
response can silently poison the cache, and every subsequent build on
that machine produces a binary from unverified bytes without any error.

There is also a loose end that belongs to the same fix:
`src/utils/toPipelineSource.mts` exists specifically to convert a web
stream into a `PipelineSource` and has its own test file, but nothing in
the package imports it — `downloadArchive.mts` suppresses the same type
error with a bare, undescribed `// @ts-expect-error` instead.

Current test coverage for `downloadArchive.mts` is 33% with lines 11-17
— the entire download body — uncovered.

## Proposed change

1. Reject non-2xx responses before touching the filesystem: check
   `response.ok` and throw an error naming the status and URL.
2. Download to a temporary path inside the cache directory, then
   `rename` it onto the final `archivePath` only after the file is
   fully written and verified. Remove the temporary file on any failure
   so a failed attempt leaves no cache entry behind.
3. Verify the archive against the official checksum: fetch
   `https://nodejs.org/dist/<version>/SHASUMS256.txt`, find the line for
   the archive filename, and compare it against the SHA-256 of the
   downloaded bytes. Fail closed — a missing or unparseable checksum
   line is an error, not a skip.
4. Replace the bare `// @ts-expect-error` with the existing
   `toPipelineSource` helper, or delete `toPipelineSource.mts` and its
   spec if the helper turns out not to fit; do not leave both the dead
   helper and the undescribed suppression in place.
5. Add tests for the new failure paths, since none of this code is
   currently exercised.

## Acceptance criteria

- A non-2xx response throws and leaves no file at the destination path.
- A response whose bytes do not match the published SHA-256 throws and
  leaves no file at the destination path.
- An interrupted write leaves no file at `archivePath`, so the next run
  re-downloads instead of reusing a partial file — covered by a test
  that simulates a mid-stream failure.
- A successful download produces the archive at `archivePath` and a
  second run reuses it without re-downloading.
- `packages/sea-builder/src/utils/downloadArchive.mts` contains no bare
  `// @ts-expect-error`, and `toPipelineSource` is either imported by
  production code or removed together with its spec.
- Statement coverage for `downloadArchive.mts` is at least 90%, and
  `pnpm run lint && pnpm run build && pnpm run test` pass.

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented._

_Effort: M -- author-estimated size; a soft autopilot selection
tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

`downloadArchive.mts` — the supply-chain download path for every SEA
binary this tool produces — performed no validation at all:

1. A non-2xx response still has a body, so a 404 page or a proxy error
   page could be streamed into the cache under the archive's name.
2. Nothing verified the downloaded bytes against `nodejs.org`'s
   published `SHASUMS256.txt`.
3. The stream wrote directly into the permanent cache path, so any
   interrupted or failed download left a file that
   `createTaskFactory`'s `existsSync` check would treat as a valid
   cache hit forever.

## Changes

- Reject non-2xx responses (both the archive fetch and the
  `SHASUMS256.txt` fetch) before touching the filesystem.
- Fetch `https://nodejs.org/dist/<version>/SHASUMS256.txt` — derived
  from the archive URL's own path, not the caller-supplied `dest`
  path, so the lookup key is correct by construction rather than by
  the two happening to match today — and verify the downloaded bytes'
  SHA-256 against it while streaming. A missing, malformed (not a
  64-character hex digest), or mismatched checksum throws; nothing is
  silently skipped, and the comparison is case-insensitive.
- Download to a unique temporary path (`pid` + a random UUID, so two
  concurrent downloads of the same target never share one — cache
  downloads run with `{ concurrent: true }` in
  `createListrCacheTasks`) and `rename` onto the final `archivePath`
  only after the checksum passes.
- `moveIntoPlace()` treats a `rename` refused by an already-placed
  destination as a benign race loss rather than an error: if two
  concurrent calls target the same `dest`, the loser's rename can fail
  (notably on Windows, which can refuse to replace a file another
  process holds open) — but since both already checksum-verified the
  same bytes, the loser just discards its own temp file instead of
  surfacing the error. A rename failure with no existing destination
  still propagates.
- Replaced the bare, undescribed `// @ts-expect-error` with a
  documented `toNodeReadable()` cast: `fetch`'s `response.body` is
  typed against the DOM lib's `ReadableStream`, which isn't assignable
  to `node:stream/promises`' `pipeline` without going through
  `Readable.fromWeb()` and a cast to `node:stream/web`'s
  `ReadableStream` type.
- **Deleted `toPipelineSource.mts` and its spec** rather than reusing
  it, per the issue's proposed alternative: it guards on `.readable`,
  a property a WHATWG `ReadableStream` does not have, so passing
  `response.body` through it would throw `TypeError` on exactly the
  input it existed to handle. After converting through
  `Readable.fromWeb()`, the guard is redundant, and nothing else in
  the package imported it.

## Verification

- Added `downloadArchive.spec.mts` (no test file existed before):
  covers a non-2xx archive response, a non-2xx `SHASUMS256.txt`
  response, a missing checksum entry, a malformed (non-hex) checksum
  entry, an uppercase-hex checksum entry (accepted), a checksum
  mismatch, a mid-stream error, and an empty body — each asserting no
  file is left at the destination path — plus the success path, a
  case where `dest`'s filename deliberately differs from the archive
  URL's filename (confirming the checksum lookup key is derived from
  the URL, not the caller-supplied destination), and a mocked-`rename`
  case covering both the benign-race-loss path (destination already
  placed) and the propagate-on-real-failure path.
- Statement coverage for `downloadArchive.mts`: **100%**, both in the
  package-local `vitest run --coverage` (vitest 4) and in the root
  aggregate `pnpm run test` report (vitest 3; was 33% before this
  change, lines 11-17 uncovered — the entire download body).
- `pnpm run lint && pnpm run build && pnpm run test` all pass (68
  tests).

## Notes for the reviewer (not defects)

- N targets on the same Node version now issue N identical
  `SHASUMS256.txt` fetches — correct, mildly redundant; caching that
  response was out of scope for this issue.
- The archive response is fetched before the `SHASUMS256.txt` fetch,
  so a `SHASUMS256.txt` failure leaves that first response's body
  unconsumed. Harmless (no file is written until after both fetches
  succeed), flagging so it doesn't read as an oversight.

Closes #57

## Summary by CodeRabbit

* **Bug Fixes**
  * Archive downloads now validate server responses and verify SHA-256 checksums before completing.
  * Failed, interrupted, or invalid downloads are cleaned up without leaving incomplete destination files.
  * Downloads are finalized atomically, improving reliability and preventing corrupted archives from being used.

* **Quality Improvements**
  * Added coverage for successful downloads, HTTP failures, missing checksums, checksum mismatches, interruptions, and empty responses.

### CODE REVIEW

File: packages/sea-builder/src/utils/downloadArchive.mts [hunk 1]
```diff
@@ -1,17 +1,100 @@
...
+ */
+const findExpectedChecksum = (shasums: string, filename: string): string => {
+  const line = shasums
+    .split('\n')
+    .find((l) => l.trim().split(/\s+/).at(-1) === filename);
+  const checksum = line?.trim().split(/\s+/).at(0);
+  if (!checksum) {
+    throw new Error(`No checksum entry found for ${filename}`);
+  }
+  return checksum;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply packages/sea-builder/src/utils/downloadArchive.mts]: Fixed in 2b869fa: `findExpectedChecksum` now validates the token against a 64-char hex regex (throwing `Malformed checksum entry` otherwise) and normalizes to lowercase before comparing.

File: packages/sea-builder/src/utils/downloadArchive.mts [hunk 2]
```diff
@@ -1,17 +1,100 @@
...
+      createWriteStream(tempPath),
+    );
+    const actualChecksum = hash.digest('hex');
+    if (actualChecksum !== expectedChecksum) {
+      throw new Error(
+        `Checksum mismatch for ${filename}: expected ${expectedChecksum}, got ${actualChecksum}`,
+      );
+    }
+    await rename(tempPath, dest);
+  } catch (error) {
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply packages/sea-builder/src/utils/downloadArchive.mts]: Fixed in 2b869fa: the temp path now includes a random UUID alongside the pid, so two concurrent downloads for the same `dest` no longer share a temp file. Also added `moveIntoPlace()`, which treats a rename refused by an already-placed destination as a benign race loss rather than an error (see reply to the CodeRabbit thread below for the Windows rename detail).

File: packages/sea-builder/src/utils/downloadArchive.mts [hunk 3]
```diff
@@ -1,17 +1,100 @@
...
+  return checksum;
+};
+
+/**
+ * Download a Node.js archive to a temporary path, verify it against the
+ * published `SHASUMS256.txt` checksum, then atomically move it onto
+ * {@link dest}. A non-2xx response, a missing or unparseable checksum
+ * entry, or a checksum mismatch all throw and leave no file behind at
+ * {@link dest}; a failed or interrupted attempt removes its own
+ * temporary file so the next run does not treat it as a cache hit.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply packages/sea-builder/src/utils/downloadArchive.mts:82]: Fixed in 2b869fa: added `moveIntoPlace()` — if `rename(tempPath, dest)` fails, it checks whether `dest` now exists (via `access`); if so, another concurrent call already placed the checksum-verified bytes there, so this call discards its own temp file instead of surfacing the EPERM/EBUSY. If `dest` doesn't exist, the original error still propagates. Added a test that mocks `rename` to reject and asserts both outcomes.
